pyuvdata/uvdata/ms.py

```diff
@@ -290,105 +1786,81 @@ def read_ms(
+
+        # Note: measurement sets use the antenna number as an index into the antenna
+        # table. This means that if the antenna numbers do not start from 0 and/or are
+        # not contiguous, empty rows are inserted into the antenna table
+        # these 'dummy' rows have positions of zero and need to be removed.
+        # (this is somewhat similar to miriad)
+        ant_good_position = np.nonzero(np.linalg.norm(full_antenna_positions, axis=1))[
+            0
+        ]
+        full_antenna_positions = full_antenna_positions[ant_good_position, :]
+        self.antenna_numbers = np.arange(n_ants_table)[ant_good_position]
+
+        # check to see if a TELESCOPE_LOCATION column is present in the observation
+        # table. This is non-standard, but inserted by pyuvdata
+        if "TELESCOPE_LOCATION" in tb_obs.colnames():
+            self.telescope_location = np.squeeze(tb_obs.getcol("TELESCOPE_LOCATION"))
+        else:
+            if self.telescope_name in self.known_telescopes():
+                # Try to get it from the known telescopes
                 self.set_telescope_params()
-            except ValueError:
+            elif xyz_telescope_frame == "ITRF":
+                # Set it to be the mean of the antenna positions (this is not ideal!)
+                self.telescope_location = np.array(
+                    np.mean(full_antenna_positions, axis=0)
+                )
+            else:
                 warnings.warn(
                     "Telescope frame is not ITRF and telescope is not "
                     "in known_telescopes, so telescope_location is not set."
                 )
+        tb_obs.close()
 
         # antenna names
-        ant_names = tb_ant.getcol("STATION")
-        ant_diams = tb_ant.getcol("DISH_DIAMETER")
-
-        self.antenna_diameters = ant_diams[ant_diams > 0]
-
-        self.Nants_telescope = len(ant_flags[np.invert(ant_flags)])
-        test_name = ant_names[0]
-        names_same = True
-        for antnum in range(len(ant_names)):
-            if not (ant_names[antnum] == test_name):
-                names_same = False
-        if not (names_same):
-            # cotter measurement sets store antenna names in the NAMES column.
+        ant_names = np.asarray(tb_ant.getcol("NAME"))[ant_good_position].tolist()
+        station_names = np.asarray(tb_ant.getcol("STATION"))[ant_good_position].tolist()
+        self.antenna_diameters = tb_ant.getcol("DISH_DIAMETER")[ant_good_position]
+
+        # importuvfits measurement sets store antenna names in the STATION column.
+        # cotter measurement sets store antenna names in the NAME column, which is
+        # inline with the MS definition doc. In that case all the station names are
+        # the same.
+        # TODO: Potentially flip which is the default -- use ant_names if filled and
+        # unique. This does seem to fail though on "day2_TDEM0003_10s_norx_1src_1spw".
+        # See the test "test_read_ms_read_uvfits".
```

Comment:
Depending on what clever words are written earlier I'd support this swap as long as we can come up with some way to catch cases like that file (which is a CASA tutorial test file I believe).

Reply:
So I did attempt this swap earlier, but it ended up throwing an error across several tests outside of `test_ms` that I couldn't immediately chase down. I haven't made a second attempt _yet_, although I could probably be talked into it.